plutus-tx/src/PlutusTx/Data/AssocList.hs

```diff
@@ -0,0 +1,436 @@
+                  Just r ->
+                    P.toBuiltinData
+                      ( These
+                          (P.unsafeFromBuiltinData v)
+                          (P.unsafeFromBuiltinData r)
+                        :: These a b
+                      )
+                  Nothing ->
+                    P.toBuiltinData (That (P.unsafeFromBuiltinData v) :: These a b)
+             in BI.mkCons (BI.mkPairData k v') (goRight tl)
+        )
+
+    res = goLeft ls `safeAppend` goRight rs
+
+    safeAppend xs1 xs2 =
+      P.matchList
+        xs1
+        xs2
+        ( \hd tl ->
+            let k = BI.fst hd
+                v = BI.snd hd
+             in insert' k v (safeAppend tl xs2)
+        )
+
+-- | Combine two 'AssocList's with the given combination function.
+unionWith ::
+  forall k a.
+  (P.UnsafeFromData a, P.ToData a) =>
+  (a -> a -> a) ->
+  AssocList k a ->
+  AssocList k a ->
+  AssocList k a
+unionWith f (toBuiltinList -> ls) (toBuiltinList -> rs) =
+  unsafeFromBuiltinList res
+  where
+    ls' :: BI.BuiltinList (BI.BuiltinPair BuiltinData BuiltinData)
+    ls' = go ls
+      where
+        go xs =
+          P.matchList
+            xs
+            nil
+            ( \hd tl ->
+                let k' = BI.fst hd
+                    v' = BI.snd hd
+                    v'' = case lookup' k' rs of
+                      Just r ->
+                        P.toBuiltinData
+                          (f (P.unsafeFromBuiltinData v') (P.unsafeFromBuiltinData r))
+                      Nothing -> v'
+                 in BI.mkCons (BI.mkPairData k' v'') (go tl)
+            )
+
+    rs' :: BI.BuiltinList (BI.BuiltinPair BuiltinData BuiltinData)
+    rs' = go rs
+      where
+        go xs =
+          P.matchList
+            xs
+            nil
+            ( \hd tl ->
+                let k' = BI.fst hd
+                    tl' = go tl
+                 in if member' k' ls
+                      then tl'
+                      else BI.mkCons hd tl'
+            )
+
+    res :: BI.BuiltinList (BI.BuiltinPair BuiltinData BuiltinData)
+    res = go rs' ls'
+      where
+        go acc xs =
+          P.matchList
+            xs
+            acc
+            (\hd -> go (BI.mkCons hd acc))
+
+{-# INLINEABLE toList #-}
+-- | Convert the `AssocList` to a list of key-value pairs. This operation is O(n).
+-- See 'toBuiltinList' for a more efficient alternative.
+toList :: (P.UnsafeFromData k, P.UnsafeFromData a) => AssocList k a -> [(k, a)]
+toList d = go (toBuiltinList d)
+  where
+    go xs =
+      P.matchList
+        xs
+        []
+        ( \hd tl ->
+            (P.unsafeFromBuiltinData (BI.fst hd), P.unsafeFromBuiltinData (BI.snd hd))
+              : go tl
+        )
+
+{-# INLINEABLE toBuiltinList #-}
+-- | Convert the `AssocList` to a `P.BuiltinList` of key-value pairs. This operation is O(1).
+toBuiltinList :: AssocList k a -> BI.BuiltinList (BI.BuiltinPair BuiltinData BuiltinData)
```

Comment:
One of the requirements here is that conversion to and from `BuiltinData` should basically be a no-op. I think if we were to use the `BuiltinList` representation then the conversions would be O(n) (see `PlutusTx.Builtins.Internal.unsafeDataAsMap`).

Reply:
Sorry, I don't understand your point. Why would it change the asymptotics, if you're calling `unsafeDataAsMap` all over the place regardless? I'm just proposing to do it earlier, at creation of `AssocList` time rather than in every function over `AssocList`.

Also, it's not O(n), even though it's defined as O(n). You can see that at the definition side it's

```haskell
{-# NOINLINE unsafeDataAsMap #-}
unsafeDataAsMap :: BuiltinData -> BuiltinList (BuiltinPair BuiltinData BuiltinData)
```

but note that `NOINLINE` pragma: it's not Plutus Tx, it's Haskell. For Plutus Tx we [replace](https://github.com/IntersectMBO/plutus/blob/6912db21d472684a06094ec09465b8767e800f98/plutus-tx-plugin/src/PlutusTx/Compiler/Builtins.hs#L406) every call to `unsafeDataAsMap` with a call to the `UnMapData` builtin, which is O(1).

Tricky, I know!